Core 6: Notifications update

ironage opened the pull request:
Upgrade to the object store PR https://github.com/realm/realm-object-store/pull/880, when merged, we should use the core-6 branch of object store again.

- ~~Removes the "oldModifications" and "newModifications" properties on the global notifier events, this might be a breaking change, depending on if these were documented or not?~~
- There is no guarantee about the ordering of keys reported by the global notifier. This was probably implied before, but now the keys are not reported in "queue" order any longer.
- Fixed behaviour of reported notifications on a clear/deleteAll event which was mostly working by accident in the previous core 6 implementation.
- Should theoretically speed up the global notifier reporting mechanism, but this is not measured.
- Notification changes now vend proper Realm Objects, instead of our internal object key representation which users couldn't do anything with.

## ☑️ ToDos
* [x] 📝 Changelog entry
* [x] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

tgoyne commented:
The typescript definitions need to be updated to reflect the removal of newModifications/oldModifications. The multi-process notifier stuff also does things with those properties; at the minimum they need to be removed from the test support code (tests/js/support/notification-worker.ts).

kneth commented:
It is a breaking change. And we need to update the [API doc](https://realm.io/docs/javascript/latest/api/Realm.Results.html#addListener) too.

ironage commented:
I took a stab at the docs updates. Looks like this was a breaking change from the typescript definitions point of view, but not from the docs point of view?

I also just realized that I'm not sure if people can actually do anything with the keys reported in this way... is there anything in js that translates a core stable id into a user object primary key? Or should we be doing this translation for them when reporting keys?

ironage commented:
@kneth @tgoyne I don't see any way the raw object ids can be used through the public API. How about we report the changes as a collection of objects; inserted and modified as live into the new realm, and deleted into the old realm? I think the translation from object-id to object can happen entirely at the js binding level.

tgoyne commented:
We probably want a way to look up objects by key and get the key from an object in the public API, but even if we do add that changing the GN API to vend objects sounds like a good idea to me.

kneth commented:
The JS api has a private method to retrieve an object using the object id. I have been waiting for a final solution on the object id to surface (to align with MongoDB's object ids).

ironage commented:
Updated the global notifier to pass out objects (there's still one failing test to figure out). This depends on getting a new core released with https://github.com/realm/realm-core/pull/3528

kneth commented on tests/js/realm-tests.js:
```diff
@@ -1529,10 +1534,15 @@ module.exports = {
+            Realm.deleteFile(config);
+        }
         Realm.copyBundledRealmFiles();
+        TestCase.assertTrue(Realm.exists(config)))
```
```suggestion
        TestCase.assertTrue(Realm.exists(config));
```
